New Resource: aws_waf_geo_match_set

MikeSouza opened the pull request:
I added a `aws_waf_geo_match_set` resource which supports the AWS WAF [GeoMatchSet](https://docs.aws.amazon.com/waf/latest/APIReference/API_GeoMatchSet.html) feature added back in October 2017.

Per the [contributing guidelines](https://github.com/terraform-providers/terraform-provider-aws/blob/master/.github/CONTRIBUTING.md), I followed the conventions used by the existing WAF resources/ match sets, included and passed acceptance tests, updated the documentation, and kept the LOC to a minimum.

I also added an acceptance test to the existing `aws_waf_rule` resource to verify that it works with the new `aws_waf_geo_match_set` resource.

This was relatively straightforward to implement, as the behavior of the new resource is identical to the existing WAF match set resources, with its main difference being a simpler schema (i.e. compared to `aws_waf_xss_match_set`).

Please let me know if you have any questions, concerns, or would like me to make any changes, thanks!

### Related
- Resolves #2529
- Fixes #2655

radeksimko commented on aws/resource_aws_waf_geo_match_set_test.go:
```diff
@@ -0,0 +1,342 @@
+
+func testAccCheckAWSWafGeoMatchSetDestroy(s *terraform.State) error {
+	for _, rs := range s.RootModule().Resources {
+		if rs.Type != "aws_waf_byte_match_set" {
```
Typo - I assume this should be `aws_waf_geo_match_set`

radeksimko commented on website/docs/r/waf_rule.html.markdown:
```diff
@@ -54,7 +54,7 @@ The following arguments are supported:
   If set to `true`, AWS WAF will allow, block, or count requests based on all IP addresses _except_ `192.0.2.44`.
 * `data_id` - (Required) A unique identifier for a predicate in the rule, such as Byte Match Set ID or IPSet ID.
-* `type` - (Required) The type of predicate in a rule, such as `ByteMatchSet` or `IPSet`
+* `type` - (Required) The type of predicate in a rule. Must be one of `IPMatch`, `ByteMatch`, `SqlInjectionMatch`, `SizeConstraint`, `XssMatch`, or `GeoMatch`.
```
Nitpick: We don't usually try to put extensive list of all valid values in our documentation, but instead link to official docs. This is to avoid a constant uphill battle of being out of date as Amazon releases new stuff and changes these things.

radeksimko commented on aws/resource_aws_waf_geo_match_set_test.go:
```diff
@@ -0,0 +1,342 @@
+
+		// Return nil if the GeoMatchSet is already destroyed
+		if awsErr, ok := err.(awserr.Error); ok {
+			if awsErr.Code() == "WAFNonexistentItemException" {
```
Do you mind using a helper function here and simplify the code that way a bit? 😉 

```go
if isAWSErr(err, "WAFNonexistentItemException", "") {
```

radeksimko commented on aws/resource_aws_waf_geo_match_set.go:
```diff
@@ -0,0 +1,201 @@
+
+	resp, err := conn.GetGeoMatchSet(params)
+	if err != nil {
+		if awsErr, ok := err.(awserr.Error); ok && awsErr.Code() == "WAFNonexistentItemException" {
```
Do you mind using a helper function here and simplify the code that way a bit? 😉 

```go
if isAWSErr(err, "WAFNonexistentItemException", "") {
```

radeksimko commented on aws/resource_aws_waf_geo_match_set.go:
```diff
@@ -0,0 +1,201 @@
+					Schema: map[string]*schema.Schema{
+						"type": {
+							Type:     schema.TypeString,
+							Optional: true,
```
This field seems to be required [per docs](https://docs.aws.amazon.com/waf/latest/APIReference/API_GeoMatchConstraint.html), do you mind reflecting it here?

radeksimko commented on aws/resource_aws_waf_geo_match_set.go:
```diff
@@ -0,0 +1,201 @@
+				Required: true,
+				ForceNew: true,
+			},
+			"geo_match_constraints": &schema.Schema{
```
As discussed in all recent WAF PRs we tend to use singular names for Sets/Lists which have nested fields of non-primitive data types, like here. This is because we expect them to be used in HCL like this:

```hcl
geo_match_constraint {
  type = "Country"
  value = "US"
}
geo_match_constraint {
  type = "Country"
  value = "CA"
}
```
as opposed to lists/sets with primitive data types, like

```hcl
countries = ["CA", "US", "CN"]
```

Do you mind changing it to `geo_match_constraint` and reflecting it in CRUD & docs too?

MikeSouza commented:
@radeksimko Thanks for reviewing radek!  I believe I've addressed all of your comments, and rebased and merged.

Unfortunately, I force-pushed the changes and so I apologize if this had caused any inconvenience or issues with reviewing/ merging.  I've learned my lesson.

bflad commented:
This has been released in [version 1.12.0 of the AWS provider](https://github.com/terraform-providers/terraform-provider-aws/blob/master/CHANGELOG.md#1120-march-23-2018). Please see the [Terraform documentation on provider versioning](https://www.terraform.io/docs/configuration/providers.html#provider-versions) or reach out if you need any assistance upgrading.
